packages/react-router/modules/useRouter.js

```diff
@@ -0,0 +1,20 @@
+import React, { useContext } from "react";
+import invariant from "tiny-invariant";
+
+import RouterContext from "./RouterContext";
+import matchPath from "./matchPath";
+
+export default function useRouter(options = {}) {
+  invariant(
+    typeof useContext === "function",
+    "The useRouter hook requires React 16.7 or greater"
+  );
+
+  let context = useContext(RouterContext);
+  let location = options.location || context.location;
+  let match = options.path
+    ? matchPath(location.pathname, options)
+    : context.match;
+
+  return { ...context, location, match };
```

Comment:
Should this be memoized?

```js
return useMemo(() => ({ location, match }), [location, match])
```

It's also probably better to not spread `context` here, just let users use `useContext` (`useRouterContext`?) for better memoization.

Reply:
Feels like premature optimization. Given that most people will just `useRouter()` and don't pass any options, it might be actually slower to handle memoization here.

Although I agree that `useRouterContext` should be exposed separately for cases when you just need access to `history/location` objects.

Reply:
I don't see any real benefit. It still needs to access the whole context in each of these and [due to various reasons](https://github.com/facebook/react/issues/14110#issuecomment-436001590) changing context value would rerender all components using that context. That would be undesired in case of `useHistory` imo.

Reply:
So that would require to make 3 different contexts then?

Reply:
Not sure what will be the final solution & implementation, but it will surely change. I recommend watching that issue. For the time being, I think we are good with `useRouter` as an experimental feature.

Reply:
I actually stopped using React router because of performance. Not everyone uses React on high end laptops. Memoization is crucial for performance.

Reply:
I'd rather not make the implementation more complex for now. If it turns out to be a problem in real apps, we can optimize it later.

As for the other hooks, their implementation is literally:

```js
function useHistory(options) {
  return useRouter(options).history;
}

function useLocation(options) {
  return useRouter(options).location;
}
```

so I don't see a lot of value in providing those up front. Besides, if people want just the `history` or `location` objects, they can always destructure `useRouter`:

```js
const { history, location } = useRouter(options);
```

Reply:
It's just that spreading the `context` in the result of `useRouter` makes removing it later for optimization a breaking change.

It's better to just expose a `useRouterContext()` to get access to the `context` while also avoiding having to copy its properties into a new object.

Reply:
Yeah, @mjackson, let's have a clean access to context there, please. It's simple and doesn't complicate anything.

```js
export default function useRouterContext(): {
  return useContext(RouterContext)
}
```

Reply:
I disagree. Context is our private API and may eventually have more stuff on it, so I'm not going to expose it. The API here matches exactly [the props that you get in `<Route>`](https://github.com/ReactTraining/react-router/blob/980bdd17e05e14fc8389c8aa39b252e9287fa15d/packages/react-router/modules/Route.js#L23). So I'm using the same thing here for consistency.

Reply:
@Kovensky It's a breaking change, but we also label it as `unstable_`.